src/net/socket_ancillary.rs

```diff
@@ -0,0 +1,777 @@
+use core::convert::TryFrom;
+use core::marker::PhantomData;
+use core::mem::{size_of, zeroed};
+use core::ptr::{self, read_unaligned};
+use core::slice;
+
+use crate::imp::c;
+use crate::imp::fd::AsFd;
+use crate::imp::syscalls::{getgid, getpid, getuid};
+use crate::io::OwnedFd;
+use crate::process::{Gid, Pid, Uid};
+
+/// Create a buffer large enough for storing some control messages as returned by `recvmsg`.
+///
+/// # Examples
+///
+/// ```
+/// # fn main() {
+/// use rustix::{cmsg_space, net::SocketCred};
+/// use rustix::io::OwnedFd;
+/// // Create a buffer big enough for a `ScmRights` message with two file descriptors.
+/// let _ = cmsg_space!([OwnedFd; 2]);
+/// // Create a buffer big enough for a `ScmRights` message and a `ScmCredentials` message.
+/// let _ = cmsg_space!(OwnedFd, SocketCred);
+/// # }
+/// ```
+#[macro_export]
+macro_rules! cmsg_space {
+    ( $( $x:ty ),* ) => {
+        {
+            [0u8; 0 $(
+                + $crate::net::CMSG_SPACE(core::mem::size_of::<$x>() as _) as usize
+            )*]
+        }
+    }
+}
+
+#[doc(hidden)]
+pub use c::CMSG_SPACE;
+
+/// TODO: document
+#[non_exhaustive]
+pub enum SendAncillaryDataUnix<'a> {
+    /// TODO: document
+    ScmRights(ScmRights<'a>),
+    /// TODO: document
+    #[cfg(any(target_os = "android", target_os = "linux",))]
+    ScmCredentials(ScmCredentials<'a>),
+}
+
+/// TODO: document
+#[non_exhaustive]
+pub enum RecvAncillaryDataUnix<'a> {
+    /// TODO: document
+    ScmRights(ScmRights<'a>),
+    /// TODO: document
+    #[cfg(any(target_os = "android", target_os = "linux",))]
+    ScmCredentials(ScmCredentials<'a>),
+}
+
+impl<'a> TryFrom<&'a c::cmsghdr> for SendAncillaryDataUnix<'a> {
+    type Error = AncillaryError;
+
+    fn try_from(cmsg: &'a c::cmsghdr) -> Result<Self, AncillaryError> {
+        unsafe {
+            let cmsg_len_zero = c::CMSG_LEN(0) as usize;
+            let data_len = (*cmsg).cmsg_len as usize - cmsg_len_zero;
+            let data = c::CMSG_DATA(cmsg).cast();
+            let data = slice::from_raw_parts(data, data_len);
+
+            match ((*cmsg).cmsg_level as _, (*cmsg).cmsg_type as _) {
+                (c::SOL_SOCKET, c::SCM_RIGHTS) => Ok(SendAncillaryDataUnix::ScmRights(ScmRights(
+                    AncillaryDataIter::new(data),
+                ))),
+                #[cfg(any(target_os = "android", target_os = "linux",))]
+                (c::SOL_SOCKET, c::SCM_CREDENTIALS) => Ok(SendAncillaryDataUnix::ScmCredentials(
+                    ScmCredentials(AncillaryDataIter::new(data)),
+                )),
+                (_, _) => Err(AncillaryError::from_cmsg(&*cmsg)),
+            }
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a c::cmsghdr> for RecvAncillaryDataUnix<'a> {
+    type Error = AncillaryError;
+
+    fn try_from(cmsg: &'a c::cmsghdr) -> Result<Self, AncillaryError> {
+        unsafe {
+            let cmsg_len_zero = c::CMSG_LEN(0) as usize;
+            let data_len = (*cmsg).cmsg_len as usize - cmsg_len_zero;
```

Comment:
Is the `*` on `*cmsg` here needed? `cmsg` here is a reference,

Reply:
fixed